File: stepik-union/src/main/java/org/stepik/core/courseFormat/StudyNode.java
package org.stepik.core.courseFormat;

import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.api.client.StepikApiClient;
import org.stepik.api.objects.StudyObject;

import java.util.List;

public interface StudyNode<D extends StudyObject, C extends StudyNode> {
    @NotNull
    String getName();

    int getPosition();

    @NotNull
    StudyStatus getStatus();

    void setStatus(@Nullable StudyStatus status);

    @NotNull
    String getDirectory();

    @NotNull
    String getPath();

    long getId();

    @Nullable
    StudyNode getParent();

    void setParent(@Nullable StudyNode parent);

    @Nullable
    StudyNode getPrevChild(@Nullable StudyNode current);

    @Nullable
    StudyNode getNextChild(@Nullable StudyNode current);

    boolean isLeaf();

    long getCourseId(@NotNull StepikApiClient stepikApiClient);

    @Nullable
    C getChildById(long id);

    @Nullable
    StudyNode<?, ?> getChildByClassAndId(@NotNull Class<? extends StudyObject> clazz, long id);

    @Nullable
    C getChildByPosition(int position);

    List<C> getChildren();

    @Nullable
    D getData();

    void setData(@Nullable D data);

    void init(@NotNull Project project, @NotNull StepikApiClient stepikApiClient, @Nullable final StudyNode parent);

    default void init(@NotNull Project project, @NotNull StepikApiClient stepikApiClient) {
        init(project, stepikApiClient, null);
    }

    void reloadData(@NotNull Project project, @NotNull StepikApiClient stepikApiClient);

    boolean getWasDeleted();

    void setWasDeleted(boolean wasDeleted);

    boolean isUnknownStatus();

    void setRawStatus(@Nullable StudyStatus status);

    void resetStatus();

    void passed();

    void setProject(@NotNull Project project);
}

File: stepik-core/src/main/kotlin/org/stepik/core/courseFormat/StudyNode.kt
package org.stepik.core.courseFormat

import com.intellij.openapi.project.Project
import org.stepik.api.client.StepikApiClient
import org.stepik.api.objects.StudyObject
import org.stepik.core.common.Loggable

interface StudyNode : Loggable {

    var project: Project?

    val id: Long

    val name: String

    val position: Int

    val directory: String

    val path: String

    var status: StudyStatus

    var parent: StudyNode?

    val isLeaf: Boolean

    val children: List<Node>

    var data: StudyObject

    var wasDeleted: Boolean

    fun getPrevChild(current: StudyNode?): StudyNode?

    fun getNextChild(current: StudyNode?): StudyNode?

    fun getCourseId(stepikApiClient: StepikApiClient): Long

    fun getChildById(id: Long): StudyNode?

    fun getChildByClassAndId(clazz: Class<out StudyNode>, id: Long): StudyNode?

    fun getChildByPosition(position: Int): StudyNode?

    fun init(project: Project?, stepikApiClient: StepikApiClient?, parent: StudyNode?)

    fun init(project: Project?, stepikApiClient: StepikApiClient?) {
        init(project, stepikApiClient, parent)
    }

    fun reloadData(project: Project, stepikApiClient: StepikApiClient)

    fun resetStatus()

    fun passed()
}
